minio/api.py

```diff
@@ -1100,64 +1101,113 @@ def get_object(self, bucket_name, object_name, offset=0, length=0,
+        :param tags: Tags for destination object.
+        :param retention: :class:`Retention` configuration object.
+        :param legal_hold: Flag to set legal hold for destination object.
+        :param metadata_directive: Directive used to handle user metadata for
+                                   destination object.
+        :param tagging_directive: Directive used to handle tags for destination
+                                   object.
         :return: :class:`ObjectWriteResult <ObjectWriteResult>` object.
 
         Example::
             minio.copy_object(
                 "my-bucketname",
                 "my-objectname",
-                "my-source-bucketname/my-source-objectname",
+                CopySource("my-source-bucketname", "my-source-objectname"),
             )
             minio.copy_object(
                 "my-bucketname",
                 "my-objectname",
-                "my-source-bucketname/my-source-objectname"
-                "?versionId=b6602757-7c9c-449b-937f-fed504d04c94",
+                CopySource(
+                    "my-source-bucketname",
+                    "my-source-objectname",
+                    version_id="b6602757-7c9c-449b-937f-fed504d04c94",
+                ),
             )
         """
         check_bucket_name(bucket_name)
         check_non_empty_string(object_name)
-        check_non_empty_string(object_source)
-        check_ssec(source_sse)
+        if not isinstance(source, CopySource):
+            raise ValueError("source must be CopySource type")
         check_sse(sse)
+        if tags is not None and not isinstance(tags, Tags):
+            raise ValueError("tags must be Tags type")
+        if retention is not None and not isinstance(retention, Retention):
+            raise ValueError("retention must be Retention type")
+        if (
+                metadata_directive is not None and
+                metadata_directive not in [COPY, REPLACE]
+        ):
+            raise ValueError(
+                "metadata directive must be {0} or {1}".format(COPY, REPLACE),
+            )
+        if (
+                tagging_directive is not None and
+                tagging_directive not in [COPY, REPLACE]
+        ):
+            raise ValueError(
+                "tagging directive must be {0} or {1}".format(COPY, REPLACE),
+            )
 
-        headers = normalize_headers(metadata)
-        if metadata:
-            headers["x-amz-metadata-directive"] = "REPLACE"
-        if conditions:
-            headers.update(conditions)
-        headers.update(source_sse.copy_headers() if source_sse else {})
-        headers.update(sse.headers() if sse else {})
-        headers['X-Amz-Copy-Source'] = quote(object_source)
+        size = -1
+        if source.offset is None and source.length is None:
+            stat = self.stat_object(
+                source.bucket_name,
+                source.object_name,
+                version_id=source.version_id,
```

Comment:
When a test is ran with `version_id` value, valid or invalid, test fails with "BadRequest".
```
from minio import Minio
from minio.commonconfig import CopySource
from datetime import datetime, timezone


# running minio erasure code
client = Minio('localhost:9001',
               access_key='minio',
               secret_key='minio123',
               secure=False)

result = client.copy_object(
    "bucket",
    "object",
    CopySource(
        "sourcebucket",
        "sourceobject",
        # version_id="ea8e1f5c-8ed4-46d5-88b4-b86faa8a5740",
        version_id="4379f791-f6a8-48f3-ada3-cce66dfafca1",
        # match_etag="8315e643ed6a5d7c9962fc0a8ef9c11f",
        # modified_since=datetime(2014, 4, 1, tzinfo=timezone.utc),
    ),
)
print(result.object_name, result.version_id)
```
$ python3.8 run-copy_object.py 
Traceback (most recent call last):
  File "run-copy_object.py", line 48, in <module>
    result = client.copy_object(
  File "/home/ersan/work/src/github.com/minio/minio-py/minio/api.py", line 1189, in copy_object
    stat = self.stat_object(
  File "/home/ersan/work/src/github.com/minio/minio-py/minio/api.py", line 1783, in stat_object
    response = self._execute(
  File "/home/ersan/work/src/github.com/minio/minio-py/minio/api.py", line 401, in _execute
    return self._url_open(
  File "/home/ersan/work/src/github.com/minio/minio-py/minio/api.py", line 385, in _url_open
    raise response_error
minio.error.S3Error: S3 operation failed; code: BadRequest, message: Bad request, resource: /sourcebucket/sourceobject, request_id: 164C54D673F7583A, host_id: None, bucket_name: sourcebucket, object_name: sourceobject

Reply:
Works for me
```py
[bala@localhost minio-py]$ PYTHONPATH=$PWD python
Python 3.7.9 (default, Aug 19 2020, 17:05:11) 
[GCC 9.3.1 20200408 (Red Hat 9.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from minio import Minio
>>> client = Minio("localhost:9000", access_key="minio", secret_key="minio123", secure=False)
>>> client.make_bucket("sourcebucket", object_lock=True)
>>> client.make_bucket("bucket", object_lock=True)
>>> import io
>>> result = client.put_object("sourcebucket", "sourceobject", io.BytesIO(b"one"), 3)
>>> result.version_id
'c19e98b4-f8eb-4ae3-89e2-a5f218eafd27'
>>> result.etag
'f97c5d29941bfb1b2fdab0874906ab82'
>>> result
<minio.helpers.ObjectWriteResult object at 0x7fe6a0912610>
>>> from minio.commonconfig import CopySource
>>> result = client.copy_object("bucket", "object", CopySource("sourcebucket", "sourceobject", version_id="c19e98b4-f8eb-4ae3-89e2-a5f218eafd27"))
>>> result.version_id
'bc6f2c2f-8976-4323-a9b6-7c59de9830b2'
>>> result.etag
'f97c5d29941bfb1b2fdab0874906ab82'
>>> res = client.get_object("sourcebucket", "sourceobject", version_id="c19e98b4-f8eb-4ae3-89e2-a5f218eafd27")
>>> res.read().decode()
'one'
>>> res = client.get_object("bucket", "object", version_id="bc6f2c2f-8976-4323-a9b6-7c59de9830b2")
>>> res.read().decode()
'one'
>>>
```